lib/IRGen/GenArchetype.cpp

```diff
@@ -374,9 +423,12 @@ const TypeInfo *TypeConverter::convertArchetypeType(ArchetypeType *archetype) {
     protocols.push_back(ProtocolEntry(protocol, impl));
   }
 
+  LayoutConstraintInfo LayoutInfo = getLayoutConstraintInfo(Type(archetype));
+
   // If the archetype is class-constrained, use a class pointer
   // representation.
-  if (archetype->requiresClass()) {
+  if (archetype->requiresClass() ||
+      LayoutInfo.isRefCountedObject()) {
```

Comment:
I think isRefCountedObject() should imply requiresClass() since we also want SIL type lowering to treat such archetypes as loadable